src/main/java/org/wildfly/security/cache/LRURealmIdentityCache.java

```diff
@@ -0,0 +1,107 @@
+package org.wildfly.security.cache;
+
+import static org.wildfly.common.Assert.checkMinimumParameter;
+
+import java.security.Principal;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ConcurrentMap;
+
+import org.wildfly.security.auth.server.RealmIdentity;
+
+/**
+ * A {@link RealmIdentityCache} implementation providing a LRU cache.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public final class LRURealmIdentityCache implements RealmIdentityCache {
+
+    /**
+     * Max active entries that are present in the cache.
+     */
+    private final int maxEntries;
+
+    /**
+     * The cached entries
+     */
+    private final ConcurrentMap<Principal, CacheEntry> cache;
+
+    /**
+     * Holds the LRU entries
+     */
+    private final ConcurrentLinkedDeque<CacheEntry> lruCache;
```

Comment:
This behavior isn't really LRU, it's least-recently-added.  And using two independent concurrent structures in any context is complex and hard to prove correct; instead how about just a synchronized LinkedHashMap subclass in LRU configuration?  That should be more than fast enough.